src/simple_key_provider.rs

```diff
@@ -11,6 +11,69 @@ use std::cell::RefCell;
 use crate::errors::{KeyDecryptionError, KeyGenerationError};
 use crate::key_provider::{DataKey, KeyProvider};
 
+// EncryptedSimpleKey relies on this size being constant. If this ever needs to be changed a new
+// version of EncryptedSimpleKey needs to be created.
+type Nonce = aes_gcm::Nonce<U12>;
+const NONCE_SIZE: usize = 12;
+
+/// A decoded intermediate representation of an encrypted simple key
+///
+/// The encoded version of the encrypted simple key looks like so:
+///
+/// | Pos  | Data                   |
+/// | -----|------------------------|
+/// | 0    | Version tag (1 byte)   |
+/// | 1-13 | Nonce       (12 bytes) |
+/// | 13-  | Encrypted key          |
+#[derive(Debug)]
+struct EncryptedSimpleKey<'a> {
+    // Keep a version tag on the key just in case the format gets changed
+    // This could have other header information - but it should be safe since we're only expected
+    // keys created by the SimpleKeyProvider.
+    version: u8,
+    nonce: &'a Nonce,
+    key: &'a [u8],
+}
+
+impl<'a> EncryptedSimpleKey<'a> {
+    /// Decode an [`EncryptedSimpleKey`] from a slice following its encoded representation
+    fn from_slice(bytes: &'a [u8]) -> Result<Self, KeyDecryptionError> {
+        if bytes.len() < 1 + NONCE_SIZE {
+            return Err(KeyDecryptionError::Other(format!(
+                "Slice was too small to load an EncryptedSimpleKey. Received: {}",
+                bytes.len()
+            )));
+        }
+
+        let version = bytes[0];
+
+        if version != 1 {
```

Comment:
If you move the version tag to the AAD, decryption will fail if the version tag is incorrect.

Reply:
Updated